ipq40xx: cellc_rtl30vw: fix imagebuilder generation

gdetal opened the pull request:
This fixes issue #10511

The image build process was modifying the generated IMAGE_KERNEL to
append rootfs information (crc). This caused:
 - sysupgrade & factory images to contain 2 times the root.squashfs
   information due to both modifying the same IMAGE_KERNEL.
 - the generated imagebuilder to contain an erroneous IMAGE_KERNEL that
   contained references to an unexisting root.squashfs (the one from
   previous cause). The RTL30VW wasn't therefore able to boot the
   generated images as they contained checksums from non existing rootfs.

This commit makes sure to use a temporary IMAGE_KERNEL to append the
rootfs information for both factory and sysupgrade images.

mpratt14 commented:
should be able to fix this without adding params to the recipe 

for example another temp file

`cp -f $(IMAGE_KERNEL) $@.kernel`

gdetal commented:
Hi @mpratt14 

I am unsure how to do that, could you expand on it ?

$(IMAGE_KERNEL) needs to stay untouched for imagebuilder to work.

mpratt14 commented:
`$@` is the build target, in this case, it expands to the full path + name of the sysupgrade.bin and factory.bin

you can append something to `$@` to create a temporary file that has the name of the image as its basename

so if `$@` expands to `/path/to/images/openwrt-target-board-sysupgrade.bin`
then `$@.kernel` expands to `/path/to/images/openwrt-target-board-sysupgrade.bin.kernel`

in your diff you added param `kernel=$$$$(IMAGE_KERNEL).factory` instead it could just be `kernel=$@.factory`

but you shouldnt even need to set those params

look at the way a similar recipe `Build/append-uImage-fakehdr` is defined here
https://github.com/openwrt/openwrt/blob/master/include/image-commands.mk
where they just make a temporary file `$@.fakehdr`

mpratt14 commented:
replace IMAGE_KERNEL with a temporary file named with build target and positional variable `$@.$1`
for factory.bin `$@.$1` expands to factory.bin.factory when you do this:

```
	IMAGE/nand-factory.bin := append-rootfshdr factory | append-ubi | qsdk-ipq-factory-nand-askey factory
```

```
define Build/append-rootfshdr
	mkimage -A $(LINUX_KARCH) \
		-O linux -T filesystem \
		-C lzma -a $(KERNEL_LOADADDR) -e $(if $(KERNEL_ENTRY),$(KERNEL_ENTRY),$(KERNEL_LOADADDR)) \
		-n root.squashfs -d $(IMAGE_ROOTFS) $@.new
	cat $(IMAGE_KERNEL) > $@.$1
	dd if=$@.new bs=64 count=1 >> $@.$1
endef
```

```
define Build/qsdk-ipq-factory-nand-askey
	$(TOPDIR)/scripts/mkits-qsdk-ipq-image.sh $@.its\
		askey_kernel $@.$1 \
		askey_fs $(IMAGE_ROOTFS) \
		ubifs $@
	PATH=$(LINUX_DIR)/scripts/dtc:$(PATH) mkimage -f $@.its $@.new
	@mv $@.new $@
endef
```

mpratt14 commented:
but because the `cat` line will overwrite with `>`, you dont even need to have different names
so you can just use `$@.kernel`

mpratt14 commented:
you may still need the param `kernel=` for `sysupgrade-tar`

but thats it

mpratt14 commented:
looking at an example elsewhere, it does need the extra $ 

`sysupgrade-tar kernel=$$$$@.kernel`

gdetal commented:
I'll update the commit based on your detailed feedback, thanks !

gdetal committed:
ipq40xx: cellc_rtl30vw: fix imagebuilder generation

This fixes #10511

The image build process was modifying the generated IMAGE_KERNEL to
append rootfs information (crc). This caused:
 - sysupgrade & factory images to contain 2 times the root.squashfs
   information due to both modifying the same IMAGE_KERNEL.
 - the generated imagebuilder to contain an erroneous IMAGE_KERNEL that
   contained references to an unexisting root.squashfs (the one from
   previous cause). The RTL30VW wasn't therefore able to boot the
   generated images as they contained checksums from non existing rootfs.

This commit makes sure to use a temporary IMAGE_KERNEL to append the
rootfs information for both factory and sysupgrade images.

Signed-off-by: Gregory Detal <gregory.detal@tessares.net>

gdetal commented:
I've tested the the updated change:

 - [x]  factory image content built from sources:

```
$ binwalk ./bin/targets/ipq40xx/generic/openwrt-ipq40xx-generic-cellc_rtl30vw-squashfs-nand-factory.bin

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             device tree image (dtb)
204           0xCC            uImage header, header size: 64 bytes, header CRC: 0x301E21B, created: 2022-09-02 06:37:23, image size: 2906964 bytes, Data Address: 0x80208000, Entry Point: 0x80208000, data CRC: 0xB91C3F77, OS: Linux, CPU: ARM, image type: OS Kernel Image, compression type: lzma, image name: "ARM OpenWrt Linux-5.10.137"
268           0x10C           device tree image (dtb)
496           0x1F0           Linux kernel ARM boot executable zImage (little-endian)
2900          0xB54           device tree image (dtb)
15616         0x3D00          xz compressed data
15847         0x3DE7          xz compressed data
2887768       0x2C1058        device tree image (dtb)
2908364       0x2C60CC        uImage header, header size: 64 bytes, header CRC: 0xAEC3FF4D, created: 2022-09-02 06:37:23, image size: 3683106 bytes, Data Address: 0x80208000, Entry Point: 0x80208000, data CRC: 0xDABC73AD, OS: Linux, CPU: ARM, image type: Filesystem Image, compression type: lzma, image name: "root.squashfs"
2908596       0x2C61B4        Squashfs filesystem, little endian, version 4.0, compression:xz, size: 3683106 bytes, 1423 inodes, blocksize: 262144 bytes, created: 2022-09-02 06:37:23
6591864       0x649578        UBI erase count header, version: 1, EC: 0x0, VID header offset: 0x800, data offset: 0x1000
```

 - [x]  factory image generated from imagebuilder (built from sources):

```
$ binwalk openwrt-imagebuilder-ipq40xx-generic.Linux-x86_64/bin/targets/ipq40xx/generic/openwrt-ipq40xx-generic-cellc_rtl30vw-squashfs-nand-factory.bin

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             device tree image (dtb)
204           0xCC            uImage header, header size: 64 bytes, header CRC: 0x301E21B, created: 2022-09-02 06:37:23, image size: 2906964 bytes, Data Address: 0x80208000, Entry Point: 0x80208000, data CRC: 0xB91C3F77, OS: Linux, CPU: ARM, image type: OS Kernel Image, compression type: lzma, image name: "ARM OpenWrt Linux-5.10.137"
268           0x10C           device tree image (dtb)
496           0x1F0           Linux kernel ARM boot executable zImage (little-endian)
2900          0xB54           device tree image (dtb)
15616         0x3D00          xz compressed data
15847         0x3DE7          xz compressed data
2887768       0x2C1058        device tree image (dtb)
2908364       0x2C60CC        uImage header, header size: 64 bytes, header CRC: 0xB5B6DA41, created: 2022-09-02 06:37:23, image size: 3682230 bytes, Data Address: 0x80208000, Entry Point: 0x80208000, data CRC: 0x2606C622, OS: Linux, CPU: ARM, image type: Filesystem Image, compression type: lzma, image name: "root.squashfs"
2908596       0x2C61B4        Squashfs filesystem, little endian, version 4.0, compression:xz, size: 3682230 bytes, 1413 inodes, blocksize: 262144 bytes, created: 2022-09-02 06:37:23
6590988       0x64920C        UBI erase count header, version: 1, EC: 0x0, VID header offset: 0x800, data offset: 0x1000
```

Both have a single `rootfs.squashfs` which is what we expect.

mpratt14 commented:
looks nice now

like I mentioned, this can now be optimized even further

notice that every positional param here are all named "kernel"

```
	IMAGE/nand-factory.bin := append-rootfshdr kernel | append-ubi | qsdk-ipq-factory-nand-askey kernel
	IMAGE/nand-sysupgrade.bin := append-rootfshdr kernel | sysupgrade-tar kernel=$$$$@.kernel | append-metadata
```

so each `$@.$1` can just be `$@.kernel`

unless you think this would be used for other parts of the image, but for now it's only for this one device's kernel

hauke commented:
Thank you for your patch, I applied it to master in 0be1b78856794589a552f662f428081600e837cc.
